wrapper/src/main/java/software/amazon/jdbc/plugin/bluegreen/BlueGreenStatusMonitor.java

```diff
@@ -0,0 +1,653 @@
+
+    this.role = role;
+    this.bgdId = bgdId;
+    this.initialHostSpec = initialHostSpec;
+    this.pluginService = pluginService;
+    this.props = props;
+    this.checkIntervalMap = checkIntervalMap;
+    this.onStatusChangeFunc = onStatusChangeFunc;
+
+    this.supportBlueGreen = (SupportBlueGreen) this.pluginService.getDialect();
+
+    executorService.submit(this::runMonitoringLoop);
+    executorService.shutdown(); // executor accepts no more tasks
+  }
+
+  protected void runMonitoringLoop() {
+    try {
+      while (!this.stop.get()) {
+        try {
+          final BlueGreenPhases oldPhase = this.currentPhase;
+          this.openConnection();
+          this.collectStatus();
+          this.collectTopology();
+          this.collectHostIpAddresses();
+          this.updateIpAddressFlags();
+
+          if ((oldPhase == null && this.currentPhase != null)
+              || (this.currentPhase != null && oldPhase != this.currentPhase)) {
+            LOGGER.finest(() -> Messages.get("bgd.statusChanged", new Object[] {this.role, this.currentPhase}));
+          }
+
+          if (this.onStatusChangeFunc != null) {
+            this.onStatusChangeFunc.onStatusChanged(
+                this.role,
+                new BlueGreenInterimStatus(
+                    this.currentPhase,
+                    this.version,
+                    this.port,
+                    this.startTopology,
+                    this.currentTopology.get(),
+                    this.startIpAddressesByHostMap,
+                    this.currentIpAddressesByHostMap,
+                    this.endpoints,
+                    this.allStartTopologyIpChanged,
+                    this.allStartTopologyEndpointsRemoved,
+                    this.allTopologyChanged));
+          }
+
+          long delayMs = checkIntervalMap.getOrDefault(
+              this.panicMode.get() ? IntervalType.HIGH : this.intervalType.get(),
+              DEFAULT_CHECK_INTERVAL_MS);
+
+          this.delay(delayMs);
+
+        } catch (InterruptedException ex) {
+          Thread.currentThread().interrupt();
+          LOGGER.finest(() -> Messages.get("bgd.interrupted", new Object[] {this.role}));
+          return;
+        } catch (Exception ex) {
+          if (LOGGER.isLoggable(Level.WARNING)) {
+            LOGGER.log(
+                Level.WARNING,
+                Messages.get("bgd.monitoringUnhandledException", new Object[] {this.role}),
+                ex);
+          }
+        }
+      }
+    } finally {
+      this.closeConnection();
+      LOGGER.finest(() -> Messages.get("bgd.threadCompleted", new Object[] {this.role}));
+    }
+  }
+
+  protected void delay(long delayMs) throws InterruptedException {
+    long start = System.nanoTime();
+    long end = start + TimeUnit.MILLISECONDS.toNanos(delayMs);
+    IntervalType currentIntervalType = this.intervalType.get();
+    boolean currentPanic = this.panicMode.get();
+    long minDelay = Math.min(delayMs, 50);
```

Comment:
What is the intention for this minDelay? Is it so we delay for **at least** 50 ms? If so I think this should be `Math.max(delayMs, 50)` instead.